### PR TITLE
Improve float32 flex attention backward tiling

### LINKED ISSUE #149767: `flex_attention` slower than manual attention implementation
### 🐛 Describe the bug

I'm not sure if this is a bug or if I am using `flex_attention` incorrectly. Also, not completely sure if flex attention is designed for such masked language modeling-style use cases. That said, I found `flex_attention` to be slower than a (compiled) manual torch implementation of attention for my use case. My actual codebase is a bit too large to post here, however, the following example contains the most critical components in a dummy form. 

I get the following results on an A100 GPU.

| Model | Estimated Runtime | Memory Usage |
|--------|--------|--------|
| `flex_attention` | 45 mins | 1150 MB |
| manual attention | 35 mins | 1300 MB |

`flex_attention` is slower, although it does save some memory.

```py
import torch
import torch.nn as nn
from torch.nn.attention.flex_attention import BlockMask, create_block_mask, flex_attention
from tqdm.auto import tqdm


def compile_flex_attention():
    try:
        print("Compling flex_attention")
        return torch.compile(flex_attention, dynamic=False)
    except Exception as e:
        print(f"Compiling flex_attention failed with error '{e}'. Retrying with mode='max-autotune'.")
        try:
            return torch.compile(flex_attention, dynamic=False, mode="max-autotune")
        except Exception as e:
            print(
                f"Compiling flex_attention failed with error: '{e}', "
                "Updating your pytorch version to nightlies may solve it, or you can set"
                "in your config dataset.packed=False to avoid using flex attention.",
            )
            raise


flex_attention_compiled = compile_flex_attention()


@torch.compiler.disable(recursive=False)
def compile_friendly_flex_attention(
    q: torch.Tensor, k: torch.Tensor, v: torch.Tensor, block_mask: BlockMask, **kwargs
) -> torch.Tensor:
    return flex_attention_compiled(q, k, v, block_mask=block_mask, **kwargs)


class Attention(nn.Module):
    def __init__(self, use_flex_attn: bool = False):
        super().__init__()
        self.d_model = 512
        self.key_value_proj_dim = 64
        self.n_heads = 8
        self.inner_dim = self.n_heads * self.key_value_proj_dim
        self.use_flex_attention = use_flex_attn

        self.q = nn.Linear(self.d_model, self.inner_dim, bias=False)
        self.k = nn.Linear(self.d_model, self.inner_dim, bias=False)
        self.v = nn.Linear(self.d_model, self.inner_dim, bias=False)
        self.o = nn.Linear(self.inner_dim, self.d_model, bias=False)

    def forward(
        self,
        hidden_states: torch.Tensor,
        mask: torch.Tensor | BlockMask = None,
        encoder_states: torch.Tensor | None = None,
    ):
        batch_size = hidden_states.shape[0]

        if encoder_states is None:
            # Self Attention
            query_states = self.q(hidden_states)
            key_states = self.k(hidden_states)
            value_states = self.v(hidden_states)
        else:
            # Cross Attention
            query_states = self.q(hidden_states)
            key_states = self.k(encoder_states)
            value_states = self.v(encoder_states)

        query_states = query_states.view(batch_size, -1, self.n_heads, self.key_value_proj_dim).transpose(1, 2)
        key_states = key_states.view(batch_size, -1, self.n_heads, self.key_value_proj_dim).transpose(1, 2)
        value_states = value_states.view(batch_size, -1, self.n_heads, self.key_value_proj_dim).transpose(1, 2)

        if self.use_flex_attention:
            assert isinstance(mask, BlockMask) or mask is None
            attn_output = compile_friendly_flex_attention(
                query_states, key_states, value_states, block_mask=mask, scale=1.0
            )
        else:
            scores = torch.matmul(query_states, key_states.transpose(3, 2))
            scores += mask
            attn_weights = nn.functional.softmax(scores, dim=-1)
            attn_output = torch.matmul(attn_weights, value_states)

        attn_output = attn_output.transpose(1, 2).contiguous().view(batch_size, -1, self.inner_dim)
        attn_output = self.o(attn_output)

        return attn_output


class Encoder(nn.Module):
    def __init__(self, use_flex_attn: bool = False):
        super().__init__()
        self.sa_layers = nn.ModuleList([Attention(use_flex_attn=use_flex_attn) for _ in range(6)])

    def forward(self, hidden_states: torch.Tensor, mask: torch.Tensor | BlockMask):
        for layer in self.sa_layers:
            hidden_states = hidden_states + layer(hidden_states, mask=mask)

        return hidden_states


class Decoder(nn.Module):
    def __init__(self, use_flex_attn: bool = False):
        super().__init__()
        self.sa_layers = nn.ModuleList([Attention(use_flex_attn=use_flex_attn) for _ in range(6)])
        self.ca_layers = nn.ModuleList([Attention(use_flex_attn=use_flex_attn) for _ in range(6)])

    def forward(
        self,
        hidden_states: torch.Tensor,
        mask: torch.Tensor | BlockMask,
        encoder_states: torch.Tensor,
        encoder_mask: torch.Tensor,
    ):
        for index in range(6):
            hidden_states = hidden_states + self.sa_layers[index](hidden_states, mask=mask)
            hidden_states = hidden_states + self.ca_layers[index](
                hidden_states, mask=encoder_mask, encoder_states=encoder_states
            )

        return hidden_states


class EncoderDecoderModel(nn.Module):
    def __init__(self, use_flex_attn: bool = False):
        super().__init__()
        self.use_flex_attn = use_flex_attn
        self.encoder = Encoder(use_flex_attn=use_flex_attn)
        self.decoder = Decoder(use_flex_attn=use_flex_attn)

    def forward(self, hidden_states: torch.Tensor, mask: torch.BoolTensor, decoder_states: torch.Tensor):
        # hidden_states: (batch_size, seq_len, d_model)
        # mask: (batch_size, seq_len)
        # decoder_states: (batch_size, 1, d_model)
        batch_size, seq_len = hidden_states.shape[:2]
        if self.use_flex_attn:

            def mask_mod(b, h, q_idx, kv_idx):
                return mask[b, kv_idx]

            encoder_sa_mask = create_block_mask(
                mask_mod, batch_size, None, mask.shape[-1], mask.shape[-1], _compile=True
            )
            decoder_sa_mask = None

            def encoder_mask_mod(b, h, q_idx, kv_idx):
                return mask[b, kv_idx]

            decoder_ca_mask = create_block_mask(encoder_mask_mod, batch_size, None, 1, seq_len, _compile=True)

        else:
            encoder_sa_mask = torch.where(mask[:, None, None, :], 0.0, float("-inf"))
            decoder_sa_mask = torch.zeros(decoder_states.shape[:-1], device=decoder_states.device)[:, None, None, :]
            decoder_ca_mask = torch.where(mask[:, None, None, :], 0.0, float("-inf"))

        encoder_states = self.encoder(hidden_states, mask=encoder_sa_mask)
        decoder_states = self.decoder(
            decoder_states, mask=decoder_sa_mask, encoder_states=encoder_states, encoder_mask=decoder_ca_mask
        )

        return decoder_states


def random_batch(batch_size, seq_len, d_model, device):
    hidden_states = torch.rand(batch_size, seq_len, d_model, device=device)
    mask = torch.rand(batch_size, seq_len, device=device) > 0.5
    decoder_states = torch.rand(batch_size, 1, d_model, device=device)
    return hidden_states, mask, decoder_states


if __name__ == "__main__":
    batch_size = 32
    num_iters = 100000
    seq_len = 128
    d_model = 512

    model = EncoderDecoderModel(use_flex_attn=True).to("cuda:0")
    model = torch.compile(model)

    for _ in tqdm(range(num_iters)):
        out = model(*random_batch(batch_size, seq_len, d_model, "cuda:0"))
        out.mean().backward()
```

I also verified that the two implementations give me the same output using:
```py
batch_size = 32
num_iters = 100000
seq_len = 128
d_model = 512

model = EncoderDecoderModel(use_flex_attn=False).to("cuda:0")
model_flex = EncoderDecoderModel(use_flex_attn=True).to("cuda:0")
model_flex.load_state_dict(model.state_dict())

model = torch.compile(model)
model_flex = torch.compile(model_flex)

batch = random_batch(batch_size, seq_len, d_model, "cuda:0")

out_torch = model(*batch)
out_flex = model_flex(*batch)

print(torch.abs(out_torch - out_flex).mean().item()) # 1.6685237369529204e-07
```

### Versions

```
Collecting environment information...
PyTorch version: 2.6.0+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.31.4
Libc version: glibc-2.35

Python version: 3.11.11 (main, Dec 11 2024, 16:28:39) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.8.0-1021-aws-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.4.131
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A100-SXM4-40GB
GPU 1: NVIDIA A100-SXM4-40GB
GPU 2: NVIDIA A100-SXM4-40GB
GPU 3: NVIDIA A100-SXM4-40GB
GPU 4: NVIDIA A100-SXM4-40GB
GPU 5: NVIDIA A100-SXM4-40GB
GPU 6: NVIDIA A100-SXM4-40GB
GPU 7: NVIDIA A100-SXM4-40GB

Nvidia driver version: 550.144.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               96
On-line CPU(s) list:                  0-95
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Xeon(R) Platinum 8275CL CPU @ 3.00GHz
CPU family:                           6
Model:                                85
Thread(s) per core:                   2
Core(s) per socket:                   24
Socket(s):                            2
Stepping:                             7
BogoMIPS:                             6000.00
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch pti fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves ida arat pku ospke
Hypervisor vendor:                    KVM
Virtualization type:                  full
L1d cache:                            1.5 MiB (48 instances)
L1i cache:                            1.5 MiB (48 instances)
L2 cache:                             48 MiB (48 instances)
L3 cache:                             71.5 MiB (2 instances)
NUMA node(s):                         2
NUMA node0 CPU(s):                    0-23,48-71
NUMA node1 CPU(s):                    24-47,72-95
Vulnerability Gather data sampling:   Unknown: Dependent on hypervisor status
Vulnerability Itlb multihit:          KVM: Mitigation: VMX unsupported
Vulnerability L1tf:                   Mitigation; PTE Inversion
Vulnerability Mds:                    Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Meltdown:               Mitigation; PTI
Vulnerability Mmio stale data:        Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Vulnerable
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Vulnerable
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Retpolines; STIBP disabled; RSB filling; PBRSB-eIBRS Not affected; BHI Retpoline
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==2.1.3
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-cusparselt-cu12==0.6.2
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] torch==2.6.0
[pip3] triton==3.2.0
[conda] numpy                     2.1.3                    pypi_0    pypi
[conda] nvidia-cublas-cu12        12.4.5.8                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.4.127                 pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.4.127                 pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.4.127                 pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.1.0.70                 pypi_0    pypi
[conda] nvidia-cufft-cu12         11.2.1.3                 pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.5.147               pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.6.1.9                 pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.3.1.170               pypi_0    pypi
[conda] nvidia-cusparselt-cu12    0.6.2                    pypi_0    pypi
[conda] nvidia-nccl-cu12          2.21.5                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.4.127                 pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.4.127                 pypi_0    pypi
[conda] torch                     2.6.0                    pypi_0    pypi
[conda] triton                    3.2.0                    pypi_0    pypi
```

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #185759

CUDA float32 flex_attention backward used the conservative 16x16 fallback tile whenever Inductor did not run an expanded autotune search. For long sequence lengths with moderate head dimensions, including the issue's S=128, D=64 shape family, that produces many small backward tiles and leaves flex_attention slower than expected.

Select 64x64 backward tiles for CUDA float32 only when replacing the exact built-in single 16x16 fallback, with static Q/KV lengths >= 128, QK/V head dims <= 128, and sparse block sizes compatible with 64-wide tiles. User bwd_* options, non-default custom choices, expanded autotune choices, ROCm, and incompatible block masks keep the existing behavior. I considered changing the global CUDA heuristic table instead, but keeping the override at the flex backward lowering point lets the guard include sequence length and sparse-block compatibility without broadening the public heuristic contract.

Fixes #149767

Generated by my agent

Benchmark Results

Command recorded in state/149767/notes. On NVIDIA B200, PyTorch 2.13.0a0+gite13b8c7, CUDA 12.8, B=32, H=8, S=128, D=64, causal flex_attention forward+backward:

before_bwd16: 0.749 ms/iter

explicit_bwd64: 0.527 ms/iter

new_default: 0.533 ms/iter

Test Plan

python test/inductor/test_flex_attention.py -k test_float32_backward_default_uses_large_tile_for_long_sequences

lintrunner -a

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says the TorchTitan failures were from optional ciflow/torchtitan CI failing during external torchao/torchcomms dependency resolution before tests ran, so I removed that label. The docs failures completed docs generation and then failed preview upload because aws was missing in the CI image, which looks unrelated to this flex_attention change.

**Review by drisspg — CHANGES_REQUESTED** [review]
I don't like how this code is structured

**Comment by jansel** [comment]
My agent says I restructured the change so the float32 long-sequence case selects a replacement FlexBwDConfig instead of writing heuristic defaults into kernel_options. I also removed the white-box helper test and kept the regression coverage as generated-code checks for the default, explicit bwd override, and incompatible block-mask cases.

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
